docs/running-calyx/fud2/rhai-api.md

```diff
@@ -0,0 +1,152 @@
+# Rhai API
+
+## High Level Rhai
+
+### defop
+
+```
+defop <op name>(<input1>: <input1 state>, <input2>: <input2 state> ...) >> <target1>: <target1 state>, <target2>, <target2 state> ... {
+    <statements>
+}
+```
+Defines an op with `<op name>` which runs `<statements>` to generate target states from input states. 
+
+The name, inputs, and targets are called the signature of the op. `<statements>` is called the body of the op.
+
+### shell
+
+```
+shell(<string>)
+```
+When called in the body of an op, that op will run `<string>` as a shell command to generate its targets. It is an error to call `shell` outside of the body of an op. Additionally, it is an error to call `shell` in the body of an op in which `shell_deps` was called prior.
+
+In the generated Ninja code, `shell` will create both a `rule` wrapping the shell command and a `build` command that invokes that rule. When a `defop` contains multiple `shell` commands, `fud2` automatically generates Ninja dependencies among the `build` command to ensure they run in order.
+
+### shell_deps
+
+```
+shell_deps(<string>, [<dep1>, <dep2>, ...], [<target1>, <target2>, ..])
+```
+When called in the body of an op, that op will run `<string>` as a shell command if it needs to generate `<target1>, <target2>, ...` from `<dep1>, <dep2>, ...`. It is an error to call `shell_deps` outside of the body of an op. Additionally, it is an error to call `shell_deps` in the body of an op in which `shell`  was called prior.
+
+Targets and deps are either strings, such as `"file1"`, or identifiers, such as if `c: calyx` existed in the signature of an op then `c` could be used as a target or dep.
+
+A call to `shell_deps` corresponds directly to a Ninja rule in the Ninja file generated by `fud2`.
+
+## Low Level Rhai
+
+A common routine for using low level Rhai is:
+1. Define states
+2. Define setup functions that register variables and rules
+3. Define an emitter function that produces commands using variables and rules created from (2)
+4. Define an op that uses setup functions and emitter to produce the target state from the start state
```

Comment:
Good overview! But maybe it would again be handy to have a complete, working Rhai script showing an end-to-end example? Maybe one that does the same thing as the above high-level Rhai script?

Reply:
I feel like the example would essentially be like what we have in the Low-Level Rhai example: https://docs.calyxir.org/running-calyx/fud2/scripts.html#example-script-in-low-level-rhai

Should I bring that example into this page and combine it with this overview, or should I have a separate example specifically for this page?

Reply:
Oh cool! I didn't realize we already had examples there. Yeah, it makes sense to me to just move everything here (i.e., to have a one-stop shop for all the Rhai scripting documentation instead of having it spread across two different pages).

Reply:
If it seems like this page is getting too long, one way to slice it would be to separate the high-level Rhai and low-level Rhai pages? That would be a logical decomposition that wouldn't require too much jumping back and forth.

Reply:
I separated out the page into General info (`rhai-api.md`), High Level Rhai (`high-level-rhai.md` - I need to write a description about high level rhai but otherwise should be good) and Low Level Rhai (`low-level-rhai.md`). If the General info page seems a bit awkward I can massage it!